content/en/blog/_posts/2024-04-02-introducing-jobset/2024-04-02-introducing-jobset.md

```diff
@@ -0,0 +1,199 @@
+complete if and only if all pods that are part of the “worker” ReplicatedJob are completed.
+
+Configurable failure policies
+: JobSet has configurable failure policies which allow the user to specify a maximum number of
+times the JobSet should be restarted in the event of a failure. If any job is marked failed,
+the entire JobSet will be recreated, allowing the workload to resume from the last checkpoint.
+When no failure policy is specified, if any job fails, the JobSet simply fails. 
+
+Exclusive placement per topology domain
+: JobSet allows users to express that child jobs have 1:1 exclusive assignment to a topology
+domain, typically an accelerator island like a rack. For example, if the JobSet creates
+two child jobs, then this feature will enforce that the pods of each child job will be
+co-located on the same island, and that only one child job is allowed to schedule per island.
+This is useful for scenarios where we want to use a distributed data parallel (DDP)
+training strategy to train a model using multiple islands of compute resources (GPU racks
+or TPU slices), running 1 model replica in each accelerator island, ensuring the forward
+and backward passes themselves occur within a single model replica occurs over the high
+bandwidth interconnect linking the accelerators chips within the island, and only the
+gradient synchronization between model replicas occurs across accelerator islands over the
+lower bandwidth data center network.
+
+Integration with Kueue
+: Users can submit JobSets via [Kueue](https://kueue.sigs.k8s.io/) to oversubscribe their clusters, queue workloads to run as capacity becomes available, prevent
+partial scheduling and deadlocks, enable multi-tenancy, and more.
+
+## Example use case
+
+### Distributed ML training on multiple TPU slices with Jax
+
+The following example is a JobSet spec for running a TPU Multislice workload on
+4 TPU v5e [slices](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#slices). To learn 
+more about TPU concepts and terminology, please refer to these [docs](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm).
+
+This example uses [Jax](https://jax.readthedocs.io/en/latest/quickstart.html), an ML framework with native support for Just-In-Time (JIT)
+compilation targeting TPU chips via [OpenXLA](https://github.com/openxla). However, you can also use 
+[PyTorch/XLA](https://pytorch.org/xla/release/2.3/index.html) to do ML training on TPUs.
+
+This example makes use of several JobSet features (both explicitly and implicitly) to support the unique scheduling requirements of
+TPU multislice training out-of-the-box with very little configuration required by the user.
+
+```yaml
+# Run a simple Jax workload on 
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: multislice
+  annotations:
+    # Give each child Job exclusive usage of a TPU slice 
+    alpha.jobset.sigs.k8s.io/exclusive-topology: cloud.google.com/gke-nodepool
```

Comment:
We should register (and, ideally, deprecate) this annotation before we publicize it.
The page to register it on is https://kubernetes.io/docs/reference/labels-annotations-taints/

---

It needs registering because although it uses a deprecated pattern (“alpha” in the key), it's inside an official Kubernetes namespace for annotations.

Reply:
If WG Batch do deprecate this annotation and switch to something different, that's a nice but separate improvement.

We still need the original one registered if any released code inside https://github.com/kubernetes/ uses it, or previously used it.

Reply:
I don't know if we should register alpha annotations/labels. Sounds like drift waiting to happen.

I don't know if most sig projects register the labels/annotations on that website.

Reply:
Marking annotations as alpha or beta is deprecated; we should just use the plain form (eg `jobset.kubernetes.io/exclusive-topology`). AIUI if a SIG wants to mark something as experimental, and for that not to be registered, it needs to be outside `kubernetes.io` or `k8s.io`.

We should register this one, _or_ pick something different and blog about the updated annotation.

Reply:
@danielvegamyhre WDYT? Could we have a post without this annotation for now? I know that changing these annotations could break users.

Reply:
Separate from this article: if / when we switch to an annotation following the recommended patterns, we can still support the old annotation as a legacy thing

[and in that case, it's still polite to register it; registering an annotation is a short 8 to 12 line PR against k/website]

Reply:
Here's an example of an annotation that's registered but not used any more
https://kubernetes.io/docs/reference/labels-annotations-taints/#service-alpha-kubernetes-io-tolerate-unready-endpoints-deprecated

Reply:
This annotation is required for the example to work. Could we just update the blog post if/when we migrate the annotation from alpha to the plain form?

Reply:
I think that in order to publish the post with those annotations, we should register them in the k8s website. 

I created https://github.com/kubernetes-sigs/jobset/issues/639 for this reason. I don't see very many sig related projects with annotations. In fact, the only project I see where annotations are registered is node feature discovery (NFD).

Reply:
That's a good way forward. It should not be much work.

Reply:
> I don't know if most sig projects register the labels/annotations on that website.

SIG Architecture don't have the capacity to follow up every place where people haven't followed the project's policy.

Reply:
Does registering include projects outside https://github.com/kubernetes/? this is a subproject not an upstream k8s api.

Reply:
Sure, but as soon as we blog about it, that article makes the annotations look official.

I do think the SIG should fix it (ie, register the annotations). _Much_ simpler than requesting a policy change.

Reply:
Imagine if a vendor, ACMECorp, had asked to publish a blog article, and they were using an unregistered annotation from the protected namespace. eg `kubernetes.io/use-acmecorp-extensions: auto`.

I'd be asking them to either change the article or change the product, for much the same reason. To me, being in-project doesn't count as an excuse for skipping the policy either.

Reply:
@ahg-g if you're wondering how much work is involved in registering the annotation, it's a 10 line PR to update https://kubernetes.io/docs/reference/labels-annotations-taints/

Please consider that if writing >= 10 lines of objection.

Reply:
Got it! The way I've done it, anything with an in-project origin can be registered (even if it shouldn't have been used; as soon as code ships that might use the annotation, we have to document it).

Reply:
Registered means “make a record that it's already in use“; it's not like an API review - deliberately!

Reply:
> https://kubernetes.io/docs/reference/labels-annotations-taints/

I can do this